src/System.ServiceModel.NetNamedPipe/src/System/ServiceModel/Channels/PipeConnectionInitiator.cs

```diff
@@ -0,0 +1,437 @@
+                                                matchPipeName = pipeName;
+                                            }
+                                        }
+                                        else
+                                        {
+                                            return pipeName;
+                                        }
+                                    }
+                                }
+                                finally
+                                {
+                                    sharedMemory.Dispose();
+                                }
+                            }
+                        }
+                        catch (AddressAccessDeniedException exception)
+                        {
+                            throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new EndpointNotFoundException(SR.Format(
+                                SR.EndpointNotFound, uri.AbsoluteUri), exception));
+                        }
+
+                        path = PipeUri.GetParentPath(path);
+                    }
+                }
+            }
+
+            if (string.IsNullOrEmpty(matchPipeName))
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                    new EndpointNotFoundException(SR.Format(SR.EndpointNotFound, uri.AbsoluteUri),
+                    new PipeException(SR.Format(SR.PipeEndpointNotFound, uri.AbsoluteUri))));
+            }
+
+            return matchPipeName;
+        }
+
+        private void PrepareConnect(Uri remoteUri, TimeSpan timeout, out string resolvedAddress, out BackoffTimeoutHelper backoffHelper)
+        {
+            PipeUri.Validate(remoteUri);
+            if (DiagnosticUtility.ShouldTraceInformation)
+            {
+                TraceUtility.TraceEvent(TraceEventType.Information, TraceCode.InitiatingNamedPipeConnection,
+                    SR.TraceCodeInitiatingNamedPipeConnection,
+                    new StringTraceRecord("Uri", remoteUri.ToString()), this, null);
+            }
+            resolvedAddress = GetPipeName(remoteUri);
+            const int backoffBufferMilliseconds = 150;
+            TimeSpan backoffTimeout;
+            if (timeout >= TimeSpan.FromMilliseconds(backoffBufferMilliseconds * 2))
+            {
+                backoffTimeout = TimeoutHelper.Add(timeout, TimeSpan.Zero - TimeSpan.FromMilliseconds(backoffBufferMilliseconds));
+            }
+            else
+            {
+                backoffTimeout = Ticks.ToTimeSpan((Ticks.FromMilliseconds(backoffBufferMilliseconds) / 2) + 1);
+            }
+
+            backoffHelper = new BackoffTimeoutHelper(backoffTimeout, TimeSpan.FromMinutes(5));
+        }
+
+        private async ValueTask<IConnection> TryConnectAsync(Uri remoteUri, string resolvedAddress, BackoffTimeoutHelper backoffHelper)
+        {
+            bool lastAttempt = backoffHelper.IsExpired();
```

Comment:
It seems odd that this method still attempts to connect after timeout

Reply:
See my previous comment, the final delay time is truncated to hit the originally specified timeout exactly. IsExpired means we just hit the deadline provided to the backoff helper. This had 150ms subtracted from it in PrepareConnect so we're actually up to 150ms before the original timeout (there's some wiggle room because of system timer resolution, so likely between 135ms and 150ms).